cpp_common/src/lib.rs

```diff
@@ -89,23 +101,67 @@ impl ClosureSig {
     }
 
     pub fn extern_name(&self) -> Ident {
-        format!("__cpp_closure_{}", self.name_hash()).into()
+        Ident::new(
+            &format!("__cpp_closure_{}", self.name_hash()),
+            Span::call_site(),
+        )
     }
 }
 
-#[derive(Clone, Debug, PartialEq, Eq)]
+#[derive(Clone, Debug)]
 pub struct Closure {
     pub sig: ClosureSig,
-    pub body: Spanned<String>,
+    pub body: TokenTree,
+    pub body_str: String, // with rust! macro replaced
     pub callback_offset: u32,
 }
 
+impl Parse for Closure {
+    fn parse(input: ParseStream) -> Result<Self> {
+        input.parse::<Option<Token![unsafe]>>()?;
```

Comment:
Can we have more comments in the parsing code explaining what each block is trying to parse from the syntax :-)?